### DISCUSSION TITLE
YARP Emulator

### DISCUSSION THREAD

**Discussion by bmatatuk** [dotnet/yarp#880] [Ideas] [answered]
There are extensive configurations possible with YARP to proxy the requests. But, I found configuration/testing can be challenging with increasing number of routes and hosts with complex match settings. It can be challenging when trying it with hostnames in the test environment. It can sometimes takes lots of trial and error approach on a server with live domains pointing to it.

**Proposal**
It would be great, with YARP extension, a .net core project can generate an emulator site for development/testing (similar to what swagger does). 
For example: When a YARP project is run, developer could access a URL http://localhost:[port]/yarp. This presents the user with a developer emulator portal. On this portal one can enter real like URL such as http://www.somedomain.com/Products or http://subdomain.domain.com/product/edit?ProductId=424. This page would display which route and/or which cluster is matched and what the proxied URL would look like. Doing this, we would virtually be able to test proxy setting without even have to host it in the live environment and also without any real domain pointing your IP.

**Comment by Tratcher** [comment, marked as answer]
It's an interesting idea. You'd be able to capture this information by submitting a real request directly to the server (with a fake host) and using either a proxy middleware or better yet, shimming out the HttpClient at the end of the pipeline to capture the outgoing proxy request after all of the transforms have run. It would short circuit and return the data to the caller in the http response. I'll experiment and see if that's viable.

**Reply by bmatatuk** [reply]
Thanks,
If you can manage to get it to work the way you said, a new sample demonstrating the concept would be beneficial to the whole community. 

Regards

**Reply by Tratcher** [reply]
How does this look? https://github.com/microsoft/reverse-proxy/commit/3bb8a28e29128163272d0a56793a1de28afbb444

Output:
```json
{
  "RouteId": "route1",
  "RouteMatch": {
    "Methods": null,
    "Hosts": null,
    "Path": "{**catch-all}",
    "Headers": null
  },
  "ClusterId": "cluster1",
  "DestinationId": "destination1",
  "DestinationPrefix": "https://localhost:10000",
  "Version": "2.0",
  "VersionPolicy": "RequestVersionOrLower",
  "Method": "GET",
  "Uri": "https://localhost:10000/prefix/",
  "RequestHeaders": [
    {
      "Key": "Cache-Control",
      "Value": [
        "max-age=0"
      ]
    },
    {
      "Key": "Accept",
      "Value": [
        "text/html",
        "application/xhtml+xml",
        "application/xml; q=0.9",
        "image/webp",
        "image/apng",
        "*/*; q=0.8",
        "application/signed-exchange; v=b3; q=0.9"
      ]
    },
    {
      "Key": "Accept-Encoding",
      "Value": [
        "gzip",
        "deflate",
        "br"
      ]
    },
    {
      "Key": "Accept-Language",
      "Value": [
        "en-US",
        "en; q=0.9"
      ]
    },
    {
      "Key": "Cookie",
      "Value": [
        ".Yarp.ReverseProxy.Affinity=CfDJ8HG3SmV%2FUfVGubjevxAy6yGu6%2FsTQgmifIeBB9a%2BqtF%2BZvHDoCZcw2jA9I0FwXSy5%2FaQtjkpjZ%2F6dEMsz2ZK%2FNoCq3splLUDspskuxujsj76GOUxDUs9VY0Vsy%2BOrKpkGw"
      ]
    },
    {
      "Key": "X-Forwarded-For",
      "Value": [
        "::1"
      ]
    },
    {
      "Key": "User-Agent",
      "Value": [
        "Mozilla/5.0",
        "(Windows NT 10.0; Win64; x64)",
        "AppleWebKit/537.36",
        "(KHTML, like Gecko)",
        "Chrome/91.0.4449.0",
        "Safari/537.36",
        "Edg/91.0.838.3"
      ]
    },
    {
      "Key": "Upgrade-Insecure-Requests",
      "Value": [
        "1"
      ]
    },
    {
      "Key": "sec-ch-ua",
      "Value": [
        "\" Not;A Brand\";v=\"99\", \"Microsoft Edge\";v=\"91\", \"Chromium\";v=\"91\""
      ]
    },
    {
      "Key": "sec-ch-ua-mobile",
      "Value": [
        "?0"
      ]
    },
    {
      "Key": "sec-fetch-site",
      "Value": [
        "none"
      ]
    },
    {
      "Key": "sec-fetch-mode",
      "Value": [
        "navigate"
      ]
    },
    {
      "Key": "sec-fetch-user",
      "Value": [
        "?1"
      ]
    },
    {
      "Key": "sec-fetch-dest",
      "Value": [
        "document"
      ]
    },
    {
      "Key": "X-Forwarded-Host",
      "Value": [
        "localhost:5001"
      ]
    },
    {
      "Key": "X-Forwarded-Proto",
      "Value": [
        "https"
      ]
    }
  ],
  "ContentHeaders": null
}
```

I didn't do the page that sends the requests, that can be done by any networking tool like Fiddler.

**Reply by samsp-msft** [reply]
What about a middleware component that looks for a custom header, and then just short circuits and reports back with the route and cluster data for where it would have been sent?

**Reply by Tratcher** [reply]
@samsp-msft That's pretty similar to what I did in https://github.com/microsoft/reverse-proxy/commit/3bb8a28e29128163272d0a56793a1de28afbb444, though I didn't make the short circuit optional. I don't think this is something you'd want to risk leaving on in production, it could leak sensitive config data.

I opted for an HttpMessageHandler instead of an ASP.NET middleware so you could observe the request after transforms had run.

**Reply by bmatatuk** [reply]
Thanks @Tratcher 

Awesome... That works for me fine. The output is comprehensive enough. 

Initially, my test app would not run at all with your code in the startup.cs (IE services.AddSingleton<IProxyHttpClientFactory, InterceptorFactory>();). After a little bit of trial and error, I have found the following line from your code and added that into my startup.cs for DI within InterceptorFactory class.
`services.AddHttpContextAccessor();`

Now my startup.cs looks like this:
`services.AddHttpContextAccessor();`
`services.AddSingleton<IProxyHttpClientFactory, InterceptorFactory>();`


Secondly, you mentioned about creating request by faking hostname in fiddler. I could not manage to get this to work with Fiddler. Instead, cURL worked for me fine. I would appreciate if you could show how to construct such a request using Fiddler (out of curiosity)

**Reply by Tratcher** [reply]
You're right, I thought I could do that with Fiddler but it keeps overwriting the host header. When testing custom hosts I usually use HttpClient. You give it the URI with the host/ip you want to connect to, but you set the Host header to a custom value. Note the custom host is also the value used for SNI in the TLS handshake.

**Reply by bmatatuk** [reply]
Looks like after preview12 something has broken this example. I would appreciate if you could fix this for me.

**Reply by Tratcher** [reply]
There were several API renames in preview12. I've updated the sample based on 1.0.0. https://github.com/microsoft/reverse-proxy/commit/9602251dcb943e081c08bf094f03528183278843

**Reply by bmatatuk** [reply]
Thanks @Tratcher.
